RESUMABLE: Remove nominative languages addressing the lost final response (#3405)
httpwg/http-extensions: draft-ietf-httpbis-resumable-upload.md

Reviewer: Should we adjust this sentence as well?

```suggestion
- `2xx (Successful)` status code, the client can continue appending representation data to it and/or mark the upload as complete ({{upload-appending}}).
```

An upload resource is specific to the upload of one representation. For uploading multiple representations, multiple upload resources have to be used.

> The server can clean up an upload resource and make it inaccessible immediately after the upload is complete. However, keeping the upload resource available for a reasonable amount of time after completion allows the client to verify the state of the upload if it did not receive the last response acknowledging the completion.

Reviewer 2: The old text said "reasonable amount of time" but I still don't know what that means, so wordsmithing a bit how about something like below (but also happy to spin this out as an editorial issue if we want)

```suggestion
The server can clean up an upload resource and make it inaccessible immediately after the upload is complete. However, keeping the upload resource available for a period of time after completion allows the client to verify the state of the upload if it did not receive the last response acknowledging the completion. Implementations are responsible for deciding if they retain the resource and for what duration; they will need to consider the resource costs required to do so.
```

An upload resource is specific to the upload of one representation. For uploading multiple representations, multiple upload resources have to be used.

> The server can clean up an upload resource and make it inaccessible immediately after the upload is complete. However, if a client didn't receive the last response acknowledging the upload's completion and the upload resource is not available anymore, the client cannot verify the upload's state with the server. Therefore, the server SHOULD keep the upload resource available for a reasonable amount of time after the upload is complete, or provide an alternate, application-specific way for the client to verify the completion of the upload.

Reviewer: I can imagine that people will ask about an upload resource's lifetime if we remove this sentence. Ideally an upload resource stays around for a brief moment after completion to allow the client to retrieve the latest state. We could trim the paragraph down tough:

> The server can clean up an upload resource and make it inaccessible immediately after the upload is complete. However, keeping the upload resource available for a reasonable amount of time after completion allows the client to verify the upload's state if it didn't receive the last response acknowledging the completion.

This refers to a HEAD request in my opinion, which is different from fetching the last response body (the motivation of this PR).

If the `Upload-Offset` request header field value does not match the current offset ({{upload-offset}}), the server MUST reject the request with a `409 (Conflict)` status code. The response MUST include the correct offset in the `Upload-Offset` header field. The response can use the problem type {{PROBLEM}} of "https://iana.org/assignments/http-problem-types#mismatching-upload-offset" ({{mismatching-offset}}).

> Even if the upload is complete ({{upload-complete}}) in the server's perspective and the final response from the targeted resource has already been sent, the client might still perform an upload append {#upload-appending} after an offset retrieval {#offset-retrieving} due to the response being lost during transmission. The server can choose to replay the final response to the client if the request to append to the completed upload is valid.

Reviewer: Can we move this paragraph down one paragraph after "If the `Upload-Complete` request header field is set to true [...]"? This other paragraph talks about generating responses, which helpful context to understand what this new sentence is about.

Author: I moved it 2 paragraphs down since they are related

If the `Upload-Complete` request header field is set to true, the client intends to transfer the remaining representation data in one request. If the request content was fully processed, the upload is marked as complete and the server SHOULD generate the response that matches what the resource, that was targeted by the initial upload creation ({{upload-creation}}), would have generated if it had processed the entire representation in the initial request. However, the response MUST include the `Upload-Complete` header field with a true value, allowing clients to identify whether a response, in particular error responses, is related to the resumable upload itself or the processing of the uploaded representation.

If the `Upload-Complete` request header field is set to false, the client intends to transfer the remaining representation data over multiple requests. If the request content was fully processed, the server acknowledges the appended data by sending a `2xx (Successful)` response with the `Upload-Complete` header field set to false.

> Even if the upload is complete ({{upload-complete}}) in the server's perspective and the final response from the targeted resource has already been sent, the client might still perform an upload append {#upload-appending} after an offset retrieval {#offset-retrieving} due to the response being lost during transmission. The server can choose to replay the final response to the client if the request to append to the completed upload is valid.

Reviewer 2: nit: I think its this syntax

```suggestion
Even if the upload is complete ({{upload-complete}}) in the server's perspective and the final response from the targeted resource has already been sent, the client might still perform an upload append {{upload-appending}} after an offset retrieval {{offset-retrieving}} due to the response being lost during transmission. The server can choose to replay the final response to the client if the request to append to the completed upload is valid.
```
